theano/tensor/nnet/abstract_conv.py

```diff
@@ -366,6 +374,215 @@ def conv2d_grad_wrt_weights(input,
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type use_1D_kernel: bool
+    :param use_1D_kernel: if set to true, row and column will
+        be upsampled seperately by 1D kernels, otherwise they are
+        upsampled together using a 2D kernel. The final result is the
+        same, only the speed can differ, given factors such as
+        upsampling ratio.
+
+    :type repeat_first_row: bool
+    :param repeat_first_row: when upsampling by an even ratio, setting
+        this variable to true copies the first row (making it appear twice
+        in the final upsampled tensor), on the other hand a false value
+        copies the last row.
+
+    :type repeat_first_col: bool
+    :param repeat_first_col: when upsampling by an even ratio, setting
+        this variable to true copies the first column (making it appear twice
+        in the final upsampled tensor), on the other hand a false value
+        copies the last column.
+
+    returns: the upsampled 4D tensor, upsampled using bilinear
+             interpolation with fixed (not learned) kernels such that
+             in the 2D space, each feature is upsampled by the given ratio.
+    """
+
+    T = theano.tensor
+    bs, ch, row, col = input.shape
+    up_input = input.reshape((-1, 1, row, col))
+
+    # indicating whether the output dimension
+    # is one dimension less (in the case of even ratio)
+    # or not (in the case of odd ratio)
+    if ratio % 2 == 0:
+        sub = 1
+    else:
+        sub = 0
+
+    # concatenating the first and last row and column
+    # first row
+    concat_mat = T.concatenate((up_input[:, :, 0:1, :], up_input), axis=2)
+    # last row
+    concat_mat = T.concatenate((concat_mat, concat_mat[:, :, -1:, :]), axis=2)
+    # first col
+    concat_mat = T.concatenate((concat_mat[:, :, :, 0:1], concat_mat), axis=3)
+    # last col
+    concat_mat = T.concatenate((concat_mat, concat_mat[:, :, :, -1:]), axis=3)
+
+    cc_bs, cc_ch, cc_row, cc_col = concat_mat.shape
+
+    pad = ratio // 2
+
+    if use_1D_kernel:
+        kern = bilinear_kernel_1D(ratio=ratio, normalize=True)
+        kern_size = kern.shape[0]
+        # upsampling rows
+        upsampled_row = conv2d_grad_wrt_inputs(output_grad=concat_mat,
+                                               filters=kern[np.newaxis,
+                                                            np.newaxis,
+                                                            :, np.newaxis],
+                                               output_grad_shape=concat_mat.
+                                               shape,
+                                               input_shape=(cc_bs, cc_ch,
+                                                            cc_row * ratio,
+                                                            cc_col),
```

Comment:
Input shape should not use `cc_row` or `cc_col`, but `row` and `col`.
There seems to be an additional issue: we would like `row` and `col` to be symbolic variables, but `conv2d_grad_wrt_inputs` needs then to be constants.